### DISCUSSION TITLE
BMI goals and binary compatibility considerations

### DISCUSSION THREAD

**Discussion by hellkite500** [csdms/bmi#139] [General]
I'm trying to understand some nuance difference in the use and application of the `interface` component of BMI and would like to get feedback from the source as well as the community.

BMI exists as a _**conceptual interface**_ described in broad terms by the [SIDL definition](https://bmi.readthedocs.io/en/stable/). At the same time, it exists as a set of supported [language definitions](https://bmi.readthedocs.io/en/stable/#:~:text=1%3A%20BMI%20languages.-,Language,Example,-C). This is what I would consider an _**application programming interface**_, or API. If this is correct, then these headers/modules represent the explicit machine/programming interface that comprises BMI _**in practice**_.

What I have seen in my experience with BMI, however, is that these two views, the conceptual interface and the API, don't always align well. For example, considering the concept of extensions as mentioned in other disscussion topics and issues, it seems that often times extensions are added to the BMI interface components (module, struct, class ect) directly to provide a "BMI Like" function that is not part of the existing specification but which provides convience and/or utility to a specific model.

This presents a compatibility and and interoperability issue, however, which seem counter to the goal of the conceptual BMI specification. My main question is which of the following goals of matter more to BMI?
- True binary compatibility and interoperability?
- Or, is the conceptual interface the primary goal and the supported BMI language definitions are more "guidelines" intended to be modified and used differently in different projects?

Perhaps it is both?  In which case I think there is some lacking documentation and communication as to how both of these are supported!

I'm happy to discuss more details on this topic, but for now I just wanted to introduce the concepts and get some input to help steer the conversation.

**Comment by RolfHut** [comment]
For the purpose of our project (eWaterCycle) it is very important that we adhere to the first: true binary compatibility and interoperability. An extension can add functionality required for a certain model / application, but should never be incompatible with the core API defined for that language.
